### PR TITLE
Fixing the config validation bug # 91

### PR DESCRIPTION
**Pull request by Prativa20**
Adding logic to check all possible keys in the resource_configuration in config file
and validate if it is a valid component name in the blueprint

Signed-off-by: Prativa Bawri <bawrip@vmware.com>

### CODE REVIEW

File: vrealize/resource.go [hunk 1]
```diff
@@ -858,38 +858,39 @@ func (vRAClient *APIClient) RequestCatalogItem(requestTemplate *CatalogItemReque
...
-			delete(resourceConfigSet, componentName)
+	var invalidKeys []string
+	// check if the keys in the resourceConfiguration map exists in the componentSet
+	// if the key in config is machine1.vsphere.custom.location, match every string after each dot
+	// until a matching string is found in componentSet.
+	// If found, it's a valid key else the component name is invalid
+	for k := range resourceConfiguration {
+		var key = k
+		var isValid bool
+		for strings.LastIndex(key, ".") != -1 {
```
Comment (raghavav) [inline vrealize/resource.go:879]: Directly assign lastIndex while doing this check. No need to evaluate twice.

File: vrealize/resource.go [hunk 2]
```diff
@@ -858,38 +858,39 @@ func (vRAClient *APIClient) RequestCatalogItem(requestTemplate *CatalogItemReque
...
+			lastIndex := strings.LastIndex(key, ".")
+			key = k[0:lastIndex]
```
Comment (raghavav) [inline vrealize/resource.go]: Should be
key = key[0:lastIndex] right?

**Review by raghavav — CHANGES_REQUESTED** [review]

File: resources/MockRequestTemplate [hunk 3]
```diff
@@ -0,0 +1,138 @@
+{
+   "type":"com.vmware.vcac.catalog.domain.request.CatalogItemProvisioningRequest",
+   "catalogItemId":"dhbh-jhdv-ghdv-dhvdd",
+   "requestedFor":"mock@mock.com",
```
Comment (raghavav) [inline resources/MockRequestTemplate:4]: Turns out 'mock.com' is a real business :)

File: vrealize/resource_test.go [hunk 4]
```diff
@@ -315,13 +309,34 @@ func TestConfigValidityFunction(t *testing.T) {
...
 
 	if err.Error() != validityErr {
 		t.Errorf("Expected: %v, but Found: %v", validityErr, err.Error())
 	}
+}
+
+// creates a mock request template from a request template template json file
+func GetMockRequestTemplate() *CatalogItemRequestTemplate {
+
+	filePath := ".." + utils.GetPathSeparator() + "resources" + utils.GetPathSeparator() + "MockRequestTemplate"
```
Comment (raghavav) [inline vrealize/resource_test.go]: There is an assumption here about the current directory? Do make sure you test this by running the test from other directories by passing the right directory to 'go test'.

File: vrealize/resource_test.go [hunk 5]
```diff
@@ -279,34 +282,25 @@ func TestChangeValueFunction(t *testing.T) {
...
+	mockConfigResourceMap["machine2.storage"] = 2
 
-	err = checkConfigValidity(&mockRequestTemplate, mockConfigResourceMap)
+	err = checkConfigValidity(mockRequestTemplate, mockConfigResourceMap)
 	if err != nil {
-		t.Errorf("The terraform config is valid. failed to validate. Expecting no error, but found %v ", err.Error())
+		t.Errorf("The terraform config is valid, failed to validate. Expecting no error, but found %v ", err.Error())
 	}
 
 	mockConfigResourceMap["mock.machine3.vSphere.mock.cpu"] = 2
```
Comment (raghavav) [inline vrealize/resource_test.go:306]: Add a comment in the code to indicate why this is a failure case and what should happen.

**Review by raghavav — CHANGES_REQUESTED** [review]
This is exactly the direction we want to go in for these tests. I have some asks, but for the most part things are good.

File: vrealize/resource.go (hunk 1 above)
Comment (raghavav) [reply vrealize/resource.go:879]: Did you try to fix this one? Were you able to get it working?

**Review by raghavav — APPROVED** [review]

File: vrealize/resource.go (hunk 1 above)
Comment (Prativa20) [reply vrealize/resource.go:879]: Tried the following, but the execution is stuck here and no error is being thrown

for lastIndex := strings.LastIndex(key, "."); lastIndex != -1; {
			key = key[0:lastIndex]
			if _, ok := componentSet[key]; ok {
				log.Info("The component name %s in the terraform config file is valid ", key)
				isValid = true
				break
			}
		}

**Pull request merged**